mysql-test/t/mysql_upgrade_file_leak.test

```diff
@@ -0,0 +1,42 @@
+-- source include/mysql_upgrade_preparation.inc
+-- source include/have_working_dns.inc
+-- source include/have_innodb.inc
+-- source include/have_partition.inc
+
+set sql_mode="";
+
+#
+# MDEV-31925 mysqld_upgrade --check-if-upgrade-is-needed leaks files
+#
+
+# Run mysql_upgrade with --check-if-upgrade-is-needed
+--echo Running mysql_upgrade with --check-if-upgrade-is-needed
+--exec $MYSQL_UPGRADE --check-if-upgrade-is-needed 2>&1
+
+# Check if temporary files related to mysql_upgrade are cleared
+--echo Checking for absence of temporary files by mysql_upgrade
+--perl
+
+# Use the temporary directory path from the MySQL configuration
+my $tmpdir = "$ENV{MYSQL_TMP_DIR}";
+
+opendir my $dir, $tmpdir or die "Cannot open directory: $tmpdir!";
+
+my @files = readdir $dir;
+closedir $dir;
+
+my $temp_file_found = 0;
+foreach my $file (@files) {
+    if ($file =~ /^mysql_upgrade-[A-Za-z0-9]+$/) {
+        $temp_file_found = 1;
+        unlink("$tmpdir/$file") or warn "Could not remove temporary file $file: $!";
+    }
+}
+die "Test failed: Found temporary file left by mysql_upgrade\n" if $temp_file_found;
```

Comment:
my perl isn't the best but could this be just `die ... if (!glob "mysql_upgrade*");`?